evm/src/coverage/visitor.rs

```diff
@@ -367,31 +360,71 @@ impl Visitor {
         }
     }
 
-    fn anchor_for(&self, loc: &ast::SourceLocation) -> ItemAnchor {
-        let instruction_counter = self
-            .source_maps
-            .get(&self.context)
-            .and_then(|source_map| {
-                source_map
-                    .iter()
-                    .enumerate()
-                    .find(|(_, element)| {
-                        element
-                            .index
-                            .and_then(|source_id| {
-                                Some(
-                                    source_id == self.source_id &&
-                                        element.offset >= loc.start &&
-                                        element.offset + element.length <=
-                                            loc.start + loc.length?,
-                                )
-                            })
-                            .unwrap_or_default()
-                    })
-                    .map(|(ic, _)| ic)
+    /// Maps a source range to a single instruction, called an anchor.
+    ///
+    /// If the anchor is executed at any point, then the source range defined by `loc` is considered
+    /// to be covered.
+    fn anchor_for(&self, loc: &ast::SourceLocation) -> eyre::Result<ItemAnchor> {
+        let source_map = self.source_maps.get(&self.context).ok_or_else(|| {
+            eyre::eyre!("could not find anchor for node: we do not have a source map")
+        })?;
+
+        let instruction = source_map
+            .iter()
+            .enumerate()
+            .find_map(|(ic, element)| {
+                if element.index? as usize == loc.index? &&
+                    element.offset >= loc.start &&
+                    element.offset + element.length <= loc.start + loc.length?
+                {
+                    return Some(ic)
+                }
+
+                None
             })
-            .unwrap_or(loc.start);
+            .ok_or_else(|| {
+                eyre::eyre!("could not find anchor: no matching instruction in range")
+            })?;
```

Comment:
This is more "brittle" in the sense that we don't just fallback, but the prior fallback made 0 sense so it's better this way